src/ctx_macro.rs

```diff
@@ -0,0 +1,239 @@
+/// ```
+/// # use rav1d::case_set;
+/// # use rav1d::src::disjoint_mut::DisjointMut;
+/// let mut buf = DisjointMut::new([0u8; 32]);
+/// let len = 16;
+/// for offset in [0, 16] {
+///     case_set!(32, len, offset, {
+///         set_disjoint!(&mut buf, 1u8);
+///     });
+/// }
+/// ```
+/// 
+/// ## Multiple buffer ranges
+/// 
+/// Multiple buffers with different lengths and offsets can be filled with the
+/// same body statements. In the following example, two buffers with different
+/// sizes are initialized by quarters.
+/// ```
+/// # use rav1d::case_set;
+/// let mut buf1 = [0u8; 32];
+/// let mut buf2 = [0u8; 64];
+/// for offset in [0, 8, 16, 24] {
+///     case_set!(16, buf=[&mut buf1[..], &mut buf2[..]], len=[8, 16], offset=[offset, offset*2], {
+///         set!(buf, len as u8 >> 3);
+///     });
+/// }
+/// ```
+/// 
+/// A more realistic example of filling multiple buffers with the same value is
+/// initializing different struct fields at the same time (from
+/// `src/decode.rs`):
+/// ```ignore
+/// case_set!(32, ctx=[(&t.l, 1), (&f.a[t.a], 0)], len=[bh4, bw4], offset=[by4, bx4], {
+///      let (dir, dir_index) = ctx;
+///      set_disjoint!(dir.seg_pred, seg_pred.into());
+///      set_disjoint!(dir.skip_mode, b.skip_mode);
+///      set_disjoint!(dir.intra, 0);
+///      set_disjoint!(dir.skip, b.skip);
+///      set_disjoint!(dir.pal_sz, 0);
+/// });
+/// ```
+/// 
+/// [`DisjointMut`]: crate::src::disjoint_mut::DisjointMut
+#[macro_export]
+macro_rules! case_set {
+    ($UP_TO:literal, $($WITH_DEFAULT:literal,)? $ctx:ident=[$($ctx_expr:expr),*], $len:ident=[$($len_expr:expr),*], $offset:ident=[$($offset_expr:expr),*], $body:block) => {
+        for ($ctx, ($len, $offset)) in core::iter::zip([$($ctx_expr,)*], core::iter::zip([$($len_expr,)*], [$($offset_expr,)*])) {
+            case_set!($UP_TO, $($WITH_DEFAULT,)? $ctx=$ctx, $len=$len, $offset=$offset, $body);
+        }
+    };
+    ($UP_TO:literal, $($WITH_DEFAULT:literal,)? $len:ident, $offset:ident, $body:block) => {
+        case_set!($UP_TO, $($WITH_DEFAULT,)? ctx=(), $len=$len, $offset=$offset, $body);
+    };
+    ($UP_TO:literal, $($WITH_DEFAULT:literal,)? $len:ident=$len_expr:expr, $offset:ident, $body:block) => {
+        case_set!($UP_TO, $($WITH_DEFAULT,)? ctx=(), $len=$len_expr, $offset=$offset, { $(($buf, $val)),* });
+    };
+    ($UP_TO:literal, $($WITH_DEFAULT:literal,)? $len:ident, $offset:ident=$offset_expr:expr, $body:block) => {
+        case_set!($UP_TO, $($WITH_DEFAULT,)? ctx=(), $len=$len, $offset=$offset_expr, $body);
+    };
+    ($UP_TO:literal, $($WITH_DEFAULT:literal,)? $len:ident=$len_expr:expr, $offset:ident=$offset_expr:expr, $body:block) => {
+        case_set!($UP_TO, $($WITH_DEFAULT,)? ctx=(), $len=$len_expr, $offset=$offset_expr, $body);
+    };
+    ($UP_TO:literal, $($WITH_DEFAULT:literal,)? $ctx:ident=$ctx_expr:expr, $len:ident=$len_expr:expr, $offset:ident=$offset_expr:expr, $body:block) => {
```

Comment:
A few nits:

1. `UP_TO` and `WITH_DEFAULT` should probably be named; it makes it a lot clearer I think, especially for optional args (`WITH_DEFAULT`).
2. `UP_TO` and `WITH_DEFAULT` can just be `:expr`s, since they can be `const` `:expr`s.  And we can use inline `const {}`s to ensure they are computed and computable at `const` time.
3. We can assign them to variables, which I think makes things easier to read.

```suggestion
    (up_to=$up_to:expr, $(with_default=$with_default:expr,)? $ctx:ident=$ctx_expr:expr, $len:ident=$len_expr:expr, $offset:ident=$offset_expr:expr, $body:block) => {
    let up_to = const { $up_to };
    let with_default: bool = const { $($with_default ||)? false };
```

Reply:
We have no instances where we need to make `$UP_TO` an expression, and I don't want to generalize this unless we need to. I don't think any of this needs to be done with my revised version that provides a separate `case_set_default!` macro.